### PR TITLE
check_palindrome (old)

### LINKED ISSUE #9: check_palindrome
# Check if a String is a Palindrome

Write a function that takes in a string and returns `True` if the string is a palindrome (reads the same backward as forward) and `False` otherwise.

### PR DESCRIPTION
**Pull request by hudaalamassi**
---
name: solution review
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/tests/test_is_palindrome.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+import unittest
+
+from solutions.is_palindrome import is_palindrome
+
+
+class TestIsPalindrome(unittest.TestCase):
+    """Test the is_palindrome function."""
+
+    def test_function_is_callable(self):
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py]: Why? I believe there is no need to test whether a function is callable. If there is any problem with calling the function, it will appear with other tests. I suggest removing lines (31-33)

File: solutions/tests/test_is_palindrome.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        actual = is_palindrome("radar")
+        self.assertEqual(actual, True)
+
+    def test_non_palindrome_string(self):
+        """it should return False if you pass a non-palindrome string."""
+        actual = is_palindrome("hello")
+        self.assertEqual(actual, False)
+
+    def test_very_long_palindrome(self):
+        """Test that a very long palindrome string is correctly identified."""
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py]: Whether an input is long or short, this doesn't affect the functionality of the code. This part might be removed too. The whole problem with this test file is that it's so long, way longer than the main code file.

File: solutions/tests/test_is_palindrome.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            "Are we not pure? “No, sir!” Panama’s moody Noriega brags. "
+            "“It is garbage!” Irony dooms a man—a prisoner up to new era."
+        )
+        self.assertEqual(actual, True)
+
+    def test_palindrome_with_heavy_punctuation(self):
+        """Test that the function return True for palindrome with heavy punctuation."""
+        actual = is_palindrome("!!!A man, a plan, a canal... Panama???!!!")
+        self.assertEqual(actual, True)
+
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py]: No need for many tests for the punctuation, one is enough, two at max! The point has been proven.

File: solutions/tests/test_is_palindrome.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        """it should return False if you pass an empty string"""
+        actual = is_palindrome("")
+        self.assertEqual(actual, False)
+
+    def test_single_space(self):
+        """it should return False if you pass a string containing only one space"""
+        actual = is_palindrome(" ")
+        self.assertEqual(actual, False)
+
+    def test_multiple_spaces(self):
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py]: This should be deleted too.

File: solutions/tests/test_is_palindrome.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        """it should return False if you pass a two-character string with different characters"""
+        actual = is_palindrome("ab")
+        self.assertEqual(actual, False)
+
+    def test_string_with_identical_punctuation(self):
+        """it should return False if you pass Two identical punctuation characters"""
+        actual = is_palindrome("!!")
+        self.assertEqual(actual, False)
+
+    def test_string_with_different_punctuation(self):
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py:79]: Punctuation was already tested, why again?

File: solutions/tests/test_is_palindrome.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        """it should return False if you pass a string containing multiple spaces"""
+        actual = is_palindrome("   ")
+        self.assertEqual(actual, False)
+
+    def test_single_character(self):
+        """it should return True if you pass a single character"""
+        actual = is_palindrome("a")
+        self.assertEqual(actual, True)
+
+    def test_two_identical_characters(self):
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py:69]: no need for this

File: solutions/tests/test_is_palindrome.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        """it should return False if you pass Two identical punctuation characters"""
+        actual = is_palindrome("!!")
+        self.assertEqual(actual, False)
+
+    def test_string_with_different_punctuation(self):
+        """it should return False if you pass strings with different punctuation marks."""
+        actual = is_palindrome("!?")
+        self.assertEqual(actual, False)
+
+    def test_case_insensitivity(self):
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py:84]: I don't believe that it's necessary to make a test case for a capital letter

File: solutions/tests/test_is_palindrome.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    def test_defensive_assertion_for_true_input(self):
+        """Test that an assertion is raised if the input is a boolean = True."""
+        with self.assertRaises(AssertionError):
+            is_palindrome(True)
+
+    def test_defensive_assertion_for_false_input(self):
+        """Test that an assertion is raised if the input is a boolean = False."""
+        with self.assertRaises(AssertionError):
+            is_palindrome(False)
+
```
Comment (byabdallah) [inline solutions/tests/test_is_palindrome.py]: When it comes to testing stuff other than strings, I believe testing one type might be enough to show that the code only accepts strings. I highly recommend shortening this test file.

File: solutions/tests/test_is_palindrome.py (hunk 1 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: I included the callability test as an initial step to ensure the function is working properly. While I know that callability will be tested later in the process, I started with it as a way of thinking: is my function first callable? I believe this approach enhances the readability of the code and can help beginners understand the process. It provides a check to confirm that the function is callable before moving on to more detailed tests. While I understand that it may seem redundant, I think it adds value for clarity. I’m open to your suggestion, but I feel it helps in making the code more beginner-friendly. what do you think?

File: solutions/tests/test_is_palindrome.py (hunk 8 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: Thanks for the feedback. I agree that testing one non-string type is sufficient, so I'll shorten the tests by removing redundant checks for other types.

File: solutions/tests/test_is_palindrome.py (hunk 1 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py]: I appreciate your thoughts, but I think calling a function is ABCs of python, so even a beginner would know how to and would know also if there is a problem.

File: solutions/tests/test_is_palindrome.py (hunk 7 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:84]: I believe testing for capital letters is essential because palindromes are case-sensitive. 

Without handling case differences (e.g., using the lower() method), the function could incorrectly return False for palindromes like "Madam" or "RaceCar."

This test ensures the function works correctly across different cases, which is crucial for accurate palindrome detection.

File: solutions/tests/test_is_palindrome.py (hunk 5 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:79]: I understand that punctuation was already tested, but I wanted to specifically address both identical and non-identical punctuation cases. 

My point was that someone might assume identical punctuation could be considered part of a palindrome, and non-identical punctuation might lead to a non-palindrome. 

However, in the case of palindromes, punctuation is ignored entirely. I believe testing both scenarios reinforces this behavior and also provides clarity for anyone reading the code.

File: solutions/tests/test_is_palindrome.py (hunk 6 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:103]: which test exactly?

File: solutions/tests/test_is_palindrome.py (hunk 2 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: Thank you for the feedback. I included the very long palindrome  and non-palindrome tests to ensure that the function can handle larger inputs without any issues. However, I understand that the core functionality doesn't depend on the length of the input, so I’ll consider removing them.

File: solutions/tests/test_is_palindrome.py (hunk 3 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: Initially, I included multiple tests for punctuation to ensure thorough coverage. In my implementation, I started with string.punctuation, but one of the tests failed while two others passed. This helped me realize the limitation of that approach, so I shifted my implementation to using unicodedata to cover a wider range of punctuation.

File: solutions/tests/test_is_palindrome.py (hunk 4 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: For the heavy punctuation test, my point of view was to test my implementation to ensure that I am covering all punctuation in my code. And that really helped me to shift my implementation.

For the empty string, I think it is essential because we need to decide what the behavior should be when the input is an empty string—should it be considered a palindrome or not? I decided that it should not be. 

The same goes for testing a string with spaces; while testing multiple spaces may not be necessary, I believe testing both a space and an empty string is essential to determine the behavior of the function.

File: solutions/tests/test_is_palindrome.py (hunk 6 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: It’s about the behavior of the function—should two characters or a single character be considered a palindrome or not? Some might treat a single character or two identical characters as a palindrome, while others may not.

I believe these are edge cases that help the coder decide how their function should behave.

File: solutions/tests/test_is_palindrome.py (hunk 1 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: I think you are right!

File: solutions/tests/test_is_palindrome.py (hunk 3 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py]: Yes I have seen the usage of unicodedata. That's why I am saying one or two is enough.

File: solutions/tests/test_is_palindrome.py (hunk 7 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py:84]: Well, the test file can be shortened by testing many thing at once. For instance, the first test checks whether the word "radar" is a palindrome. It can be written as "RaDaR" to check case insensitivity.

File: solutions/tests/test_is_palindrome.py (hunk 6 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py:69]: These are not edge cases I believe. Edge cases are when the input is for instance a string or null undefined values. Anyone who knows the concept of palindrome will know that "aa" is indeed a palindrome. The same applies on "a" because it's one letter only. I am still thinking there is no need for this. Generally, test file should test normal scenario of normal occurrence and edge cases (when the function **might** have a different behavior). Will entering "a" or "aa" as an input cause something unusual to happen? What is that? and Why? Try to condense the number of tests you have in this file.

File: solutions/tests/test_is_palindrome.py (hunk 5 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py:79]: I would prefer merging this test with another one to save some space and make it easy for the reader

File: solutions/tests/test_is_palindrome.py (hunk 3 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py]: I kept the additional tests to demonstrate the development process of the strategy and implementation. However, you might be right, and I can simplify them. I’ll remove one of the tests that initially passed.

File: solutions/tests/test_is_palindrome.py (hunk 5 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:79]: I want the behavior of my code to be clear when someone reads the tests, and to clear up any wrong assumptions about the behavior. I think keeping the tests for identical and non-identical punctuation is important to highlight how punctuation is handled in palindrome checks. However, I agree that tests for like heavy punctuation or other phrases may be excessive. I’ll remove those additional tests to keep the focus on the core behavior without losing clarity.

File: solutions/tests/test_is_palindrome.py (hunk 6 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: Some might consider a single character or two identical characters as non-palindromes because they aren't actual words. I can for example consider them non-palindromes in my code because I see them as not an actual words, so this need clarification in my opinion.

Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: I think these scenarios should be taken into account. We should not assume that everyone has the same understanding of palindromes. Maybe I want my program to behave differently?

File: solutions/tests/test_is_palindrome.py (hunk 7 above)
Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:84]: I prefer to keep the test cases separate because I want to clarify one point at a time. First, I want to ensure that my function correctly detects simple palindromes and behaves as expected in those cases. Then, I want to specifically address how my function handles capital letters and case sensitivity. By keeping these tests separate, I can isolate each behavior and ensure my function works correctly in both scenarios. This approach helped me implement and refine the code more strategically, addressing specific concerns step by step.

File: solutions/tests/test_is_palindrome.py (hunk 6 above)
Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py:69]: Yes you can, but you didn't, which makes them normal palindromes, not edge cases. I just noticed too, why testing "ab"?

Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: Yes. I didn't but others may have different opinion, so I should indecate that I want to treat them as palindrome. Also testing "ab" applies for the same princible. someone may srtucture their code that all "aa" and "ab" not palindromes because they are not actual words(that is their strategy and how their function will behave). But I include it, to say ( my strategy and how I want my function to behave is that "aa" is palindrom and"ab" is not palindrome and that the reason behined not considering "ab" as a palindrome is that it is not the same backword and forward not beacuse it is "not an actual word").

Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: For me, I treated 'a' and 'aa' as palindromes, but my view is that I should test them just to show how my function behaves for other developers, not to test exactly.

Testing those cases demonstrates that I am assuming that not everyone shares the same interpretation of a palindrome.

Comment (byabdallah) [reply solutions/tests/test_is_palindrome.py:69]: I believe testing "ab" is redundant as no way it can be considered as a palindrome.

Comment (hudaalamassi) [reply solutions/tests/test_is_palindrome.py:69]: I realize that my point for including this test might not have been fully understood. Perhaps you’re right, and there’s no real need for it after all. Thank you for the feedback—I'll revisit this and adjust accordingly.

**Review by byabdallah — DISMISSED** [review]

**Review by byabdallah — DISMISSED** [review]

**Review by byabdallah — APPROVED** [review]

**Review by byabdallah — DISMISSED** [review]

**Review by byabdallah — DISMISSED** [review]

**Review by byabdallah — APPROVED** [review]

**Comment by hudaalamassi** [comment]
- The merge was reverted to fulfill this rule "The function's strategy is not described in any docstrings or **tests**". 

- A new pull request will address the necessary changes. Please review and provide feedback.
